cannon/mipsevm/witness.go

```diff
@@ -49,16 +49,14 @@ func (wit *StepWitness) EncodePreimageOracleInput() ([]byte, error) {
 
 	switch preimage.KeyType(wit.PreimageKey[0]) {
 	case preimage.LocalKeyType:
-		// We have no on-chain form of preparing the bootstrap pre-images onchain yet.
-		// So instead we cheat them in.
-		// In production usage there should be an on-chain contract that exposes this,
-		// rather than going through the global keccak256 oracle.
+		if len(wit.PreimageValue) > 32+8 {
+			return nil, fmt.Errorf("local pre-image exceeds maximum size of 32 bytes with key 0x%x", wit.PreimageKey)
+		}
 		var input []byte
-		input = append(input, CheatBytes4...)
-		input = append(input, uint32ToBytes32(wit.PreimageOffset)...)
+		input = append(input, LoadLocalDataBytes4...)
 		input = append(input, wit.PreimageKey[:]...)
 		var tmp [32]byte
-		copy(tmp[:], wit.PreimageValue[wit.PreimageOffset:])
+		copy(tmp[:], wit.PreimageValue[8:])
```

Comment:
This doesn't work right now because the `PreimageOracle` loads only 1 or 2 parts at offsets 0 and 32 respectively, depending on the pre-image length.
Since the VM can load at most 4 bytes of the pre-image from memory, the op-program may request pre-image parts at any offset greater than zero.

Reply:
Should be working now.